sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -82,7 +82,8 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
   def inferRowType(rowSoFar: Array[DataType], next: Array[String]): Array[DataType] = {
     var i = 0
     while (i < math.min(rowSoFar.length, next.length)) {  // May have columns on right missing.
-      rowSoFar(i) = inferField(rowSoFar(i), next(i))
+      val typeElemInfer = inferField(rowSoFar(i), next(i))
+      rowSoFar(i) = compatibleType(rowSoFar(i), typeElemInfer).getOrElse(StringType)
```

Comment:
shall we do this inside `inferField`?

Reply:
Ok it's more intuitive and inferField is only used here and in the tests